### ISSUE TITLE
Top-scope warning when looping though an array of hashes

### ISSUE THREAD

**Issue by rorybrowne** [rodjek/puppet-lint#464]
Hello

When using .each on an array of hashes, I get an error that the variable I'm using for the current item in that array is top-scoped, if I attempt to retrieve an item from the hash. I have found a workaround involving an intermediate variable, which means I'm not directly accessing elements within the hash named by the each.

I hope this excerpt from my shell will make this clear:

``` shell
$ cat test.pp
# Test 1
class test {
  $things = [
    {
      verb    => 'hello',
      subject => 'world',
    },
  ]

  # function call with lambda:
  $things.each |$thing| {
    notify { $thing['verb'] : }
  }
}


$ cat test2.pp
# test 2
class test {
  $things = [
    {
      verb    => 'hello',
      subject => 'world',
    },
  ]

  # function call with lambda:
  $things.each |$arg_thing| {
    $thing = $arg_thing
    notify { $thing['verb'] : }
  }
}


$ puppet-lint --no-autoloader_layout-check /tmp/test.pp
WARNING: top-scope variable being used without an explicit namespace on line 12
$
$ puppet-lint --no-autoloader_layout-check /tmp/test2.pp
$
```

**Comment by qralston** [comment]
We also noticed this problem. @rorybrowne, thanks for the trick of using a transient intermediate variable.

This is a pretty big deal for us, because we've just started using Puppet 4 language constructs, and we also have pre-commit checks that reject commits that generate puppet-lint warnings or errors.

The trick of using a transient intermediate variable works, but that's a very ugly way to work around the problem.

@rodjek, any thoughts on how to fix this the right way?

**Comment by rendhalver** [comment]
I am seeing the same thing in a couple of internal modules as well.

**Comment by bitfield** [comment]
Similarly:

	  $mymodule::repos.each |Hash $repo| {
	    $reponame = $repo['name']

gives

		WARNING: top-scope variable being used without an explicit namespace

**Comment by igalic** [comment]
my code:
```puppet
class profiles::htpasswd {
  $htpasswds = hiera_hash('profiles::htpasswds', {})
  $htpasswds.each |$target, $user_pass| {
    $user_pass.each |$user, $password| {
      htpasswd { $user:
        target   => $target,
        password => $password, }}}}
```
my warnings
```
WARNING: top-scope variable being used without an explicit namespace on line 4
WARNING: top-scope variable being used without an explicit namespace on line 6
```

**Comment by rnelson0** [comment]
This was resolved with the batch of changes in #636, #637, #638:

```
$ cat 464.pp
class test {
  $things = [
    {
      verb    => 'hello',
      subject => 'world',
    },
  ]

  # function call with lambda:
  $things.each |$thing| {
    notify { $thing['verb'] : }
  }
}

$mymodule::repos.each |Hash $repo| {
  $reponame = $repo['name']

class profiles::htpasswd {
  $htpasswds = hiera_hash('profiles::htpasswds', {})
  $htpasswds.each |$target, $user_pass| {
    $user_pass.each |$user, $password| {
      htpasswd { $user:
        target   => $target,
        password => $password,
      }
    }
  }
}
$ be puppet-lint 464.pp
ERROR: test not in autoload module layout on line 1
ERROR: profiles::htpasswd not in autoload module layout on line 18
WARNING: class not documented on line 1
WARNING: class not documented on line 18
```
